src/classes/CW_AudienceGroupTriggerHandler.cls

```diff
@@ -0,0 +1,219 @@
+		List<CW_Audience_Group_Schedule__c> agss = new List<CW_Audience_Group_Schedule__c>();
+		for(CW_Audience_Group__c gr : groupsMap.values())
+		{
+			for(CW_Schedule__c sch : campSchedules.get(gr.Campaign__c))
+			{
+				CW_Audience_Group_Schedule__c ags = new CW_Audience_Group_Schedule__c();
+				ags.Audience_Group__c = gr.id;
+				ags.Schedule__c = sch.id;
+				ags.Is_Control_Group__c = gr.Is_Control_Group__c;
+				agss.add(ags);
+			}
+		}
+		insert agss;
+		
+		// part2: match control group schedules
+		matchControlGroups(agss,groupsMap);
+	}
+	
+	// fix links to control groups
+	public static void matchControlGroups(List<CW_Audience_Group_Schedule__c> agss, Map<Id,CW_Audience_Group__c> groupsMap)
+	{
+		
+		Map<String,CW_Audience_Group_Schedule__c> allAGS = createComboSet(agss);
+		
+		Set<Id> missingIds = new Set<Id>();	// control groups schedules we will need to select from the database
+		for(CW_Audience_Group__c gr : groupsMap.values())
+		{
+			if(gr.Control_Group__c != null && !groupsMap.containsKey(gr.Control_Group__c))	// if the control group is not in trigger list
+			{
+				missingIds.add(gr.Control_Group__c);
+			}
+		}
+		
+		if(!missingIds.isEmpty())	// find any that might be missing from the trigger list
+		{
+			for(CW_Audience_Group_Schedule__c ags : 
+					[select id, Audience_Group__c, Schedule__c from CW_Audience_Group_Schedule__c where Audience_Group__c in :missingIds])
+			{
+				allAGS.put(makeComboKey(ags.Audience_Group__c,ags.Schedule__c),ags);
+			}
+		}
+		missingIds.clear();
+		Set<String> removeKeys = new Set<String>();
+		
+		for(CW_Audience_Group_Schedule__c ags : allAGS.values())
+		{
+			if(groupsMap.containsKey(ags.Audience_Group__c))	// to avoid additionally added ags-s
+			{
+				Id controlGroup = groupsMap.get(ags.Audience_Group__c).Control_Group__c;
+				if(String.isNotBlank(controlGroup))
+				{
+					String comboKey = makeComboKey(controlGroup,ags.Schedule__c);
+					if(allAGS.containsKey(comboKey))
+					{
+						ags.Control_Group_Schedule__c = allAGS.get(comboKey).Id;
+					}
+				}
+			}
+			else
+			{
+				// we can remove AGS which weren't passed in
+				removeKeys.add(makeComboKey(ags.Audience_Group__c,ags.Schedule__c));
+			}
+		}
+		
+		for(String comboKey : removeKeys)	// no need to update AGS which were added from missingIds
+		{
+			allAGS.remove(comboKey);
+		}
+		
+		update allAGS.values();	// i'm updating all AGS that were passed in to make method more reusable
+	}
+	
+	// this should happen on before delete because the required lookup will otherwise prevent delete
+	public static void cascadeDelete(List<CW_Audience_Group__c> groups)
+	{
+		List<CW_Audience_Group_Schedule__c> removeAGS = new List<CW_Audience_Group_Schedule__c>();
+		List<CW_Audience_Group_Schedule_Reward__c> removeAGSR = new List<CW_Audience_Group_Schedule_Reward__c>();
+		
+		for(CW_Audience_Group_Schedule__c ags : 
+			[select id,(select id from Audience_Groups_Schedule_Reward__r) 
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.